doc/source/reference/arrays.promotion.rst

```diff
@@ -6,90 +6,104 @@
   np.int8(2)
 
-While mixing two different ``dtypes`` normally results in the larger of both:
+Mixing two different dtypes normally produces a result with the dtype of the
+higher precision input:
 
-  >>> np.int8(4) + np.int64(8)
+  >>> np.int8(4) + np.int64(8)  # 64 > 8
   np.int64(12)
-  >>> np.int16(3) + np.float64(3)
-  np.float64(6.0)
-
-In the majority of cases, this should not lead to surprises.
-However, especially if you work with non-default datatypes like low
-precision integers or floats or unsigned integers there are some details
-of NumPy promotion rules which may be relevant to know.
-These detailed rules also do not always match those of many other languages.[#hist-reasons]_
-
-For Numerical values, you can think of promotion happening in three kinds:
-``unsigned integers < signed integers < float < complex``,
-where the result will always be the highest kind of any of the inputs.
-Further, the result will always have a precision higher or equivalent than
-any of the inputs which leads to two some examples which may be unexpected:
-1. When mixing floating point numbers and integers, the integer ``dtype`` may
-   force the result to a higher precision floating point.
-2. When mixing unsigned and signed integers, the result will be a higher
-   precision than both inputs.  And an unfortunate surprise is that
-   ``int64`` and ``uint64`` can return floating point values.
+  >>> np.float32(3) + np.float16(3)  # 32 > 16
+  np.float32(6.0)
+
+In typical cases, this does not lead to surprises. However, if you work with
+non-default dtypes like unsigned integers and low-precision floats, or if you
+mix NumPy integers, NumPy floats, and Python scalars, some
+details of NumPy promotion rules may be relevant. Note that these detailed
+rules do not always match those of other languages [#hist-reasons]_.
+
+Numerical dtypes come in four "kinds" with a natural hierarchy.
+
+1. unsigned integers (``uint``)
+2. signed integers (``int``)
+3. float (``float``)
+4. complex (``complex``)
+
+In addition to kind, NumPy dtypes also have an associated precision, specified
+in bits. Together, the kind and precision specify the dtype. For example, a
+``uint8`` is an unsigned integer stored using 8 bits.
+
+The result of an operation will always be of an equal or higher kind of any of
+the inputs. Furthermore, the result will always have a precision greater than
+or equal to those of the inputs. Already, this can lead to some examples which
+may be unexpected:
+
+1. When mixing floating point numbers and integers, the precision of the
+   integer may force the result to a higher precision floating point. For
+   example, the result of an operation involving ``int64`` and ``float16``
+   is ``float64``.
+2. When mixing unsigned and signed integers with the same precision, the
+   result will have a *higher* precision than either inputs. For instance,
+   the result of an operation involving ``int64`` and ``uint64`` is
+   ``float64``.
```

Comment:
```suggestion
   result will has *higher* precision than either inputs. Additionally,
   if one of them has 64bit precision already, no higher precision integer
   is available and for example an operation involving ``int64`` and ``uint64``
   gives ``float64``.
```
This got the intention slightly off, I think.  I would love to point out that this is a known caveat and we might consider changing it, but...